### PR TITLE
Refactored get_commands(), additional tests

### PR DESCRIPTION
**Pull request by charles-cowart**

### CODE REVIEW

**Comment by charles-cowart** [comment]
Some additional tests. generate_commands() rewritten to be more flexible and support differing products from amplicon and meta*omic pipelines.

**Review by wasade — CHANGES_REQUESTED** [review]

File: qp_klp/Step.py [hunk 1]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+        blanks_files = [x for x in listdir(self.pipeline.output_path) if
+                        x.endswith('_blanks.tsv')]
+
+        uploads_folders = {x[2]: x[1] for x in self.special_map}
+
+        cmds = []
+        for blanks_file in blanks_files:
+            # assume filename like:
+            # 180716_M05314_0106_000000000-BYB7J_Metcalf_NIJ_14989_blanks.tsv
+            qiita_id = blanks_file.split('_')[-2]
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:360]: Yup. In Pipeline.generate_sample_info_files():
https://github.com/biocore/mg-scripts/blob/860d5d274cebb4f8744143b7896f50689d0b047c/sequence_processing_pipeline/Pipeline.py#L306

File: qp_klp/Step.py [hunk 2]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+        pf_files = [x for x in listdir(pf_dir) if x.endswith('.tsv')]
+
+        cmds = []
+
+        uploads_folders = {x[2]: x[1] for x in self.special_map}
+
+        for pf_file in pf_files:
+            # assume filename like:
+            # 0000-L446B.CMI_LOreal_Acne_16SV1-V3_14901.1.tsv
+            qiita_id = pf_file.split('_')[-1].split('.')[0]
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:382]: Yes. In the seqpro command: https://github.com/charles-cowart/metagenomics_pooling_notebook/blob/439586c4d6d6ee8b4e35d13fcbff75a18adfa287/metapool/scripts/seqpro.py#L61

File: qp_klp/Step.py [hunk 3]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+            uploads_folder = uploads_folders[qiita_id]
+
+            if not exists(uploads_folder):
+                raise ValueError(f"Uploads folder '{uploads_folder}' does "
+                                 "not exist")
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:387]: I don't imagine so. It's really just me being defensive; it's not a condition where I'd feel comfortable silently ignoring it and continue processing. It also has value as a sanity check on the test environment - making sure one was created for the project.

File: qp_klp/Step.py [hunk 4]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+
+        return cmds
+
+    def _helper_process_fastq_files(self):
+        cmds = []
+
+        data_dirs = [x for x in
+                     listdir(join(self.pipeline.output_path, 'QCJob')) if
+                     x in ['amplicon', 'filtered_sequences',
+                           'trimmed_sequences']]
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:399]: Can't argue with more defensive. :)

File: qp_klp/Step.py [hunk 5]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+    def generate_commands(self):
+        cmds = []
+
+        if exists(join(self.pipeline.output_path, 'ConvertJob/logs')):
+            cmds.append('tar zcvf logs-ConvertJob.tgz ConvertJob/logs')
+
+        if exists(join(self.pipeline.output_path, 'ConvertJob/Reports')):
+            cmds.append('tar zcvf reports-ConvertJob.tgz ConvertJob/Reports '
+                        'ConvertJob/Logs')
+        elif exists(join(self.pipeline.output_path, 'ConvertJob/Logs')):
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:430]: Yeah. Right now bcl2fastq is used for amplicon runs while bcl-convert is used for meta*omic ones. The latter generates a Reports directory while the former does not. We want them both in the reports-ConvertJob.tgz tarball if both are present hence the elif. It's slightly confusing with the separate 'log' directories that capture the stdout and stderr for these jobs; they go into a separate logs-ConvertJob.tgz.

File: qp_klp/Step.py [hunk 6]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+            # assume paths like:
+            # QCJob/Finrisk_12142/filtered_sequences
+            # QCJob/ABTX_11052/amplicon
+            tmp, _ = split(data_dir)
+            _, tmp = split(tmp)
+            _, qiita_id = tmp.split('_')
+
+            uploads_folder = uploads_folders[qiita_id]
+
+            if not exists(uploads_folder):
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:413]: definitely!

File: qp_klp/Step.py [hunk 7]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+        if exists(join(self.pipeline.output_path, 'ConvertJob/Reports')):
+            cmds.append('tar zcvf reports-ConvertJob.tgz ConvertJob/Reports '
+                        'ConvertJob/Logs')
+        elif exists(join(self.pipeline.output_path, 'ConvertJob/Logs')):
+            cmds.append('tar zcvf reports-ConvertJob.tgz ConvertJob/Logs')
+
+        if exists(join(self.pipeline.output_path, 'FastQCJob/logs')):
+            cmds.append('tar zcvf logs-FastQCJob.tgz FastQCJob/logs')
+
+        if exists(join(self.pipeline.output_path, 'FastQCJob/logs')):
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:436]: Thanks for the catch!

File: qp_klp/Step.py [hunk 8]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+
+        cmds.append(self._helper_process_blanks())
+
+        # if one or more tar-gzip files are found (which we expect there to
+        # be), move them into the 'final_results' directory.
+        results = [x for x in listdir(self.pipeline.output_path) if
+                   x.endswith('.tgz')]
+
+        if len(results) > 0:
+            'mv *.tgz final_results'
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:464]: Right. Cmds is executed after the list has been built up. I think the confusion here is that results implies that something has already executed. I've changed it to 'tarballs' to better describe the results of this search.

File: qp_klp/Step.py [hunk 9]
```diff
@@ -637,6 +780,32 @@ def update_blanks_in_qiita(self, qclient):
 
                 return data
 
+    def precheck(self, qclient):
+        # compare sample-ids/tube-ids in sample-sheet/mapping file
+        # against what's in Qiita.
+        results = self._compare_samples_against_qiita(qclient)
+
+        if results is not None:
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:788]: Nope, it's actually a good thing.

File: qp_klp/tests/test_metagenomic_step.py [hunk 10]
```diff
@@ -48,15 +49,64 @@ def test_metagenomic_creation(self):
...
         step.convert_bcl_to_fastq()
 
     def test_metagenomic_quality_control(self):
-        self._delete_test_output()
         self._create_test_input(2)
 
         step = Metagenomic(self.pipeline, self.qiita_id, None)
         step.quality_control()
+
+    def atest_generate_commands(self):
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/tests/test_metagenomic_step.py:63]: Thanks! That's actually vestigial code to be removed.

File: qp_klp/Step.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:464]: Nvm, I see what you're saying. Working on an alternate solution.

File: qp_klp/Step.py [hunk 11]
```diff
@@ -326,23 +326,167 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+                                 "not exist")
+
+            cmds.append("mv {data_dir}/*  {uploads_folder}")
+
+        return cmds
+
+    def generate_commands(self):
+        cmds = []
+
+        if exists(join(self.pipeline.output_path, 'ConvertJob/logs')):
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py]: I ran with this idea - let me know what you think.

File: qp_klp/Amplicon.py [hunk 12]
```diff
@@ -93,15 +93,22 @@ def generate_reports(self):
 
     def _get_data_type(self, prep_file_path):
         metadata = Step.parse_prep_file(prep_file_path)
-        if 'target_gene' in metadata.columns:
+        target_genes = []
+        for sample_name in metadata:
+            row = metadata[sample_name]
+            if 'target_gene' in row:
+                target_genes.append(row['target_gene'])
```
Comment (antgonza) [inline qp_klp/Amplicon.py:100]: Why not use pandas methods for this?
```
if 'target_gene' in metadata.columns:
   target_genes = metadata.target_gene.unique()
else:
   target_gene = []

File: qp_klp/Step.py [hunk 13]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+
+        if len(results) > 0:
+            return 'tar zcvf sample-files.tgz' + ' ' + ' '.join(results)
+
+    def _helper_process_blank_files(self):
+        blanks_files = [a_file for a_file in listdir(self.pipeline.output_path)
+                        if a_file.endswith('_blanks.tsv')]
+
+        # where proj[2] is a Qiita ID and proj[1] is a project uploads path.
+        uploads_folders = {proj[2]: proj[1] for proj in self.special_map}
```
Comment (antgonza) [inline qp_klp/Step.py:362]: We are no longer uploading anything to the uploads_folder; then, do we still need this variable?

File: qp_klp/Step.py [hunk 14]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+        cmds = []
+
+        valid_dir_names = [join(self.pipeline.output_path, 'QCJob', a_dir) for
+                           a_dir in ['amplicon', 'filtered_sequences',
+                                     'trimmed_sequences']]
+
+        data_dirs = [a_dir for a_dir in valid_dir_names if
+                     exists(a_dir) and isdir(a_dir)]
+
+        uploads_folders = {proj[2]: proj[1] for proj in self.special_map}
```
Comment (antgonza) [inline qp_klp/Step.py:385]: ... if no need to upload files, do we need this?

File: qp_klp/Step.py [hunk 15]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+                raise ValueError(f"Unexpected directory structure: {data_dir}")
+
+            qiita_id = match.groups()[0]
+            uploads_folder = uploads_folders[qiita_id]
+            cmds.append(f"mv {data_dir}/*  {uploads_folder}")
+
+        return cmds
+
+    def _helper_process_operations(self):
+        MOVE_CMD = 'mv'
```
Comment (antgonza) [inline qp_klp/Step.py:404]: are there other mv commands?

File: qp_klp/Step.py [hunk 16]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+            # do not add the command to the list unless at least one of
+            # the inputs exists. It's okay for a command to go unprocessed.
+            if confirmed_inputs:
+                # convert to string form before using.
+                confirmed_inputs = ' '.join(confirmed_inputs)
+                if order == 'AOI':
+                    cmds.append(f'{action} {output} {confirmed_inputs}')
+                elif order == 'AIO':
+                    cmds.append(f'{action} {confirmed_inputs} {output}')
+                else:
```
Comment (antgonza) [inline qp_klp/Step.py:466]: Does this level of abstraction is needed? Either way, why?

File: qp_klp/Step.py [hunk 17]
```diff
@@ -356,42 +522,61 @@ def _generate_commands(self):
...
+                        raise ValueError(f"Unrecognized file: {fastq_file}")
+
+                files['raw_barcodes'].sort()
+                files['raw_forward_seqs'].sort()
+                files['raw_reverse_seqs'].sort()
             else:
                 raise PipelineError("QCJob output not in expected location")
+
+            # TODO: need a test to check that files isn't empty and if it is
+            # raise an error instead of failing silently!
```
Comment (antgonza) [inline qp_klp/Step.py:569]: Still needed?

File: qp_klp/Step.py [hunk 18]
```diff
@@ -635,7 +821,42 @@ def update_blanks_in_qiita(self, qclient):
...
+                            f" include: {examples_in_qiita}")
+
+                if uses_tids:
+                    msgs.append(f"Project '{p_name}' is using tube-ids. You "
+                                "may be using sample names in your file.")
+
+            if msgs:
+                raise PipelineError('\n'.join(msgs))
+
+        # Note: falling through and returning None is normal and expected
```
Comment (antgonza) [inline qp_klp/Step.py:858]: Is it?

File: qp_klp/tests/test_metagenomic_step.py [hunk 19]
```diff
@@ -48,14 +47,12 @@ def test_metagenomic_creation(self):
         self.assertIsNotNone(step)
 
     def test_metagenomic_convert_bcl_to_fastq(self):
-        self._delete_test_output()
```
Comment (antgonza) [inline qp_klp/tests/test_metagenomic_step.py:51]: Out of curiosity, why aren't these needed?

File: qp_klp/tests/test_step.py [hunk 20]
```diff
@@ -331,7 +409,6 @@ def test_creation(self):
         self.assertIsNotNone(step)
 
     def test_convert_bcl_to_fastq(self):
-        self._delete_test_output()
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:334]: Why aren't these needed?

File: qp_klp/tests/test_step.py [hunk 21]
```diff
@@ -414,6 +489,9 @@ def test_generate_pipeline(self):
 
         self.assertIsNotNone(pipeline)
 
+        # don't construct a fake pipeline type to test Error raise on an
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:492]: Is this comment needed?

File: qp_klp/tests/test_step.py [hunk 22]
```diff
@@ -602,3 +680,155 @@ def test_get_tube_ids_from_qiita(self):
...
+        self.assertEqual(results[2]['project_name'], 'Gerwick')
+
+        self.assertEqual(results[0]['tids'], True)
+        self.assertEqual(results[1]['tids'], True)
+        self.assertEqual(results[2]['tids'], False)
+
+        # 'EP448041B04' is a sample-name from the sample-sheet and should not
+        # be in fake-Qiita, as defined in FakeQiita() class. Therefore, it
+        # should appear in the 'samples_not_in_qiita' list.
+        self.assertTrue('EP448041B04' in results[0]['samples_not_in_qiita'])
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:711]: You could use in self.assertIsIn, no?

File: qp_klp/tests/test_step.py [hunk 23]
```diff
@@ -602,3 +680,155 @@ def test_get_tube_ids_from_qiita(self):
...
+             'uploads/6123'),
+            (f'cd {self.output_file_path}; (find *.tgz -maxdepth 1 -type f '
+             '| xargs mv -t final_results) || true')]
+
+        # replace unique string w/the base-directory path in the expected
+        # output.
+        for i in range(0, len(exp)):
+            exp[i] = exp[i].replace('BASE_DIRECTORY', getcwd())
+
+        self.assertEqual(set(step.cmds), set(exp))
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:794]: Why do they need to be converted to a set? It should test that they are the same, no?

**Review by antgonza — CHANGES_REQUESTED** [review]
A few comments.

File: qp_klp/tests/test_metagenomic_step.py (hunk 19 above)
Comment (charles-cowart) [reply qp_klp/tests/test_metagenomic_step.py:51]: The base test class now contains a proper tearDown() method.

Comment (charles-cowart) [reply qp_klp/tests/test_metagenomic_step.py:51]: see above.

File: qp_klp/tests/test_step.py (hunk 20 above)
Comment (charles-cowart) [reply qp_klp/tests/test_step.py:334]: see above.

File: qp_klp/Amplicon.py (hunk 12 above)
Comment (charles-cowart) [reply qp_klp/Amplicon.py:100]: It's because metadata is a dict(), not a DataFrame. I'll include an option to return a DataFrame instead and modify accordingly.

File: qp_klp/Step.py (hunk 13 above)
Comment (charles-cowart) [reply qp_klp/Step.py:362]: Actually we're still moving blanks files to the uploads folder:
https://github.com/charles-cowart/qp-knight-lab-processing/blob/535cf6b28f6194051aade2e1b34c4cceb17bf9be/qp_klp/Step.py#L371

File: qp_klp/Step.py (hunk 14 above)
Comment (charles-cowart) [reply qp_klp/Step.py:385]: I believe so. We still use this code to help move the filtered fastq files to the uploads directory:
https://github.com/charles-cowart/qp-knight-lab-processing/blob/535cf6b28f6194051aade2e1b34c4cceb17bf9be/qp_klp/Step.py#L399

File: qp_klp/Step.py [hunk 24]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+            uploads_folder = uploads_folders[qiita_id]
+            cmds.append(f"mv {data_dir}/*  {uploads_folder}")
+
+        return cmds
+
+    def _helper_process_operations(self):
+        MOVE_CMD = 'mv'
+        RESULTS_DIR = 'final_results'
+        ORDER_AOI = 'AOI'   # for Action, Output, Input(s)
+        ORDER_AIO = 'AIO'   # for Action, Input(s), Output
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:407]: tar commands are of the form 'tar output.tar dir1 dir2 file1 file2...' while mv commands are of the form 'mv src1 src2 dst' - this allows us to reuse the loops for both tarballing and moves, without duplicating it. I'm open to suggestions though.

File: qp_klp/Step.py (hunk 13 above)
Comment (antgonza) [reply qp_klp/Step.py:362]: Well, we shouldn't; can you remove?

File: qp_klp/Step.py (hunk 14 above)
Comment (antgonza) [reply qp_klp/Step.py:385]: We shouldn't, please remove

File: qp_klp/Step.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (antgonza) [reply qp_klp/Step.py:407]: Right but not sure why we need the abstraction if the code can have the command directly in the system call or command ... not sure, I follow the need

File: qp_klp/Step.py (hunk 13 above)
Comment (charles-cowart) [reply qp_klp/Step.py:362]: Sure! Uh, we'll just move them to the final_results folder instead? Or should we not generate them anymore?

File: qp_klp/tests/test_step.py (hunk 23 above)
Comment (charles-cowart) [reply qp_klp/tests/test_step.py:794]: Good catch. Previously the order of some individual operations was undetermined from macosx to linux because of (I believe) walk(). As it's now fixed in order, we can test the order as well.

**Comment by charles-cowart** [comment]
@antgonza ready for review

File: qp_klp/Step.py [hunk 25]
```diff
@@ -137,7 +137,7 @@ def update_sample_sheet(cls, sample_sheet_path, lane_number):
             sheet.write(f)
 
     @classmethod
-    def parse_prep_file(cls, prep_file_path):
+    def parse_prep_file(cls, prep_file_path, convert_to_dict=True):
```
Comment (antgonza) [inline qp_klp/Step.py:140]: Just to be sure, is this really needed? I mean is there calls to parse_prep_file where convert_to_dict is True? I only see Falses here.

File: qp_klp/Step.py [hunk 26]
```diff
@@ -326,23 +328,134 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+            return []
+
+    def _helper_process_blanks(self):
+        results = [x for x in listdir(self.pipeline.output_path) if
+                   x.endswith('_blanks.tsv')]
+
+        if len(results) > 0:
+            return 'tar zcvf sample-files.tgz' + ' ' + ' '.join(results)
+
+    def _helper_process_operations(self):
```
Comment (antgonza) [inline qp_klp/Step.py:357]: I still don't understand the complexity of this method; it basically creates a list of commands that are abstracted via multiple variables and arrays; I mean, why not simply have:

```python
commands = [
    # convert logs
    'tar zcvf logs-ConvertJob-tgz ConvertJob/logs',
     ...
     ...
]

```

File: qp_klp/Step.py [hunk 27]
```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+        if result:
+            cmds.append(result)
+
+        cmds.append(self._helper_process_blanks())
+
+        # if one or more tar-gzip files are found (which we expect there to
+        # be), move them into the 'final_results' directory. However, if none
+        # are present, don't raise an error.
+        cmds.append('(find *.tgz -maxdepth 1 -type f | xargs mv -t '
+                    'final_results) || true')
```
[the remark this answers is not in the document]
Comment (antgonza) [reply qp_klp/Step.py:437]: Still not sure why?

File: qp_klp/Step.py [hunk 28]
```diff
@@ -326,23 +328,134 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
...
+
+        # confirm uploads files exist as expected before using helper
+        # methods.
+        for uploads_folder in [proj[1] for proj in self.special_map]:
+            if not exists(uploads_folder):
+                raise ValueError(f"Uploads folder '{uploads_folder}' does "
+                                 "not exist")
```
Comment (antgonza) [inline qp_klp/Step.py:444]: If nothing is gonna be uploaded, do we still need this?

File: qp_klp/Step.py [hunk 29]
```diff
@@ -356,42 +469,64 @@ def _generate_commands(self):
...
+                    else:
+                        raise ValueError(f"Unrecognized file: {fastq_file}")
+
+                files['raw_barcodes'].sort()
+                files['raw_forward_seqs'].sort()
+                files['raw_reverse_seqs'].sort()
             else:
                 raise PipelineError("QCJob output not in expected location")
+
+            for f_type in files:
```
Comment (antgonza) [inline qp_klp/Step.py:515]: :+1:

File: qp_klp/Step.py (hunk 18 above)
Comment (antgonza) [reply qp_klp/Step.py:808]: Do we still need this Note?

File: qp_klp/tests/test_step.py [hunk 30]
```diff
@@ -541,6 +616,11 @@ def test_parse_prep_file(self):
 
         self.assertDictEqual(obs, exp)
 
+        # simply confirm that a DataFrame is returned when convert_to_dict is
+        # False. We already know that the contents of obs will be correct.
+        obs = Step.parse_prep_file(good_prep_file, convert_to_dict=False)
+        self.assertTrue(isinstance(obs, pd.DataFrame))
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:622]: you can use assertIsInstance

File: qp_klp/tests/test_step.py [hunk 31]
```diff
@@ -602,3 +682,143 @@ def test_get_tube_ids_from_qiita(self):
...
 
         self.assertDictEqual(obs, exp)
+
+    def test_compare_samples_against_qiita(self):
+        fake_client = FakeClient()
+        step = Step(self.pipeline, self.qiita_id, None)
+        results = step._compare_samples_against_qiita(fake_client)
+
+        # confirm projects in results match what's expected
+        obs = set([x['project_name'] for x in results])
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:692]: why do you have to reiterate and convert to a set to test? Shouldn't you be testing the output of the function without any transformations?

File: qp_klp/tests/test_step.py [hunk 32]
```diff
@@ -602,3 +682,143 @@ def test_get_tube_ids_from_qiita(self):
...
+
+        # confirm projects using tube-ids match what's expected
+
+        # results are a list of project dicts, rather than a dict of dicts.
+        # however they are faked and can be expected to be returned in a
+        # fixed order. Assert the order is as expected so the following tests
+        # will be meaningful.
+        self.assertEqual(results[0]['project_name'], 'NYU_BMS_Melanoma')
+        self.assertEqual(results[1]['project_name'], 'Feist')
+        self.assertEqual(results[2]['project_name'], 'Gerwick')
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:704]: You could use assertCountEqual - since python3 it also checks values.

File: qp_klp/tests/test_step.py [hunk 33]
```diff
@@ -602,3 +682,143 @@ def test_get_tube_ids_from_qiita(self):
...
+
+        # 'EP448041B04' is a sample-name from the sample-sheet and should not
+        # be in fake-Qiita, as defined in FakeQiita() class. Therefore, it
+        # should appear in the 'samples_not_in_qiita' list.
+        self.assertIn('EP448041B04', results[0]['samples_not_in_qiita'])
+
+        # 'BLANK3.3B' is defined in the sample-sheet and also in FakeQiita,
+        # both as a sample-name and as a tube-id (One of the few to be so
+        # named). It shouldn't appear in 'samples_not_in_qiita' list.
+        self.assertTrue('BLANK3.3B' not in results[0]['samples_not_in_qiita'])
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:718]: assertNotIn

File: qp_klp/tests/test_step.py [hunk 34]
```diff
@@ -602,3 +682,143 @@ def test_get_tube_ids_from_qiita(self):
...
+
+        self.assertEqual(step.cmds, exp)
+
+    def test_overwrite_prep_files(self):
+        # use a prep-file specifically for modification by the
+        # _overwrite_prep_files() method.
+        fake_client = FakeClient()
+        step = Step(self.pipeline, self.qiita_id, None)
+
+        # 20220423_FS10001773_12_BRB11603-0615.Matrix_Tube_LBM_14332.1.tsv
```
Comment (antgonza) [inline qp_klp/tests/test_step.py:792]: Do you need this comment?

**Review by antgonza — COMMENTED** [review]
Several comments.

File: qp_klp/Step.py (hunk 25 above)
Comment (charles-cowart) [reply qp_klp/Step.py:140]: Arguably the most important use of parse_prep_file():
https://github.com/charles-cowart/qp-knight-lab-processing/blob/37736e223f6f6f7ed10ec985974ad1ec09467df8/qp_klp/Step.py#L188

File: qp_klp/Step.py (hunk 26 above)
Comment (charles-cowart) [reply qp_klp/Step.py:357]: This was what we ran with originally. Daniel suggested we generalize it:
https://github.com/qiita-spots/qp-knight-lab-processing/pull/67#discussion_r1223255298

File: qp_klp/Step.py (hunk 28 above)
Comment (charles-cowart) [reply qp_klp/Step.py:444]: fair point. :)

File: qp_klp/Step.py (hunk 16 above)
Comment (charles-cowart) [reply qp_klp/Step.py]: Some actions like 'tar' require defining the output before the inputs while others like 'mv' require all inputs to be given before the output. At any rate it's been simplified some.

File: qp_klp/Step.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (charles-cowart) [reply qp_klp/Step.py:437]: I also would prefer a python-based solution but at the time this command occurred in the middle of a series of commands.

File: qp_klp/tests/test_step.py (hunk 31 above)
Comment (charles-cowart) [reply qp_klp/tests/test_step.py:692]: That's only true because the call to Qiita is faked and the order is fixed, but I get you.

**Comment by charles-cowart** [comment]
@antgonza ready!
